File: sleep_timer_feature/src/main/java/com/parabola/sleep_timer_feature/SleepTimerInteractor.java
package com.parabola.sleep_timer_feature;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;

public interface SleepTimerInteractor {

    Completable start(long timeToSleepMs);

    Completable reset();

    boolean launched();

    Single<Long> remainingTimeToEnd();

    Observable<Long> observeRemainingTimeToEnd();

    Observable<Boolean> observeIsTimerRunning();

    Observable<Integer> onTimerFinished();


    class TimerNotLaunchedException extends IllegalStateException {
    }

    class TimerAlreadyLaunchedException extends IllegalStateException {
    }
}

File: sleep_timer_feature/src/main/java/com/parabola/sleep_timer_feature/SleepTimerInteractor.kt
package com.parabola.sleep_timer_feature

import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single

interface SleepTimerInteractor {
    fun start(timeToSleepMs: Long): Completable

    fun reset(): Completable

    fun launched(): Boolean

    fun remainingTimeToEnd(): Single<Long>

    fun observeRemainingTimeToEnd(): Observable<Long>
    fun observeIsTimerRunning(): Observable<Boolean>

    fun onTimerFinished(): Observable<Int>

    class TimerNotLaunchedException : IllegalStateException()
    class TimerAlreadyLaunchedException : IllegalStateException()
}
